File: php/sword_offer_32_i.php
<?php

// 剑指 Offer 32 - I. 从上到下打印二叉树

// 从上到下打印出二叉树的每个节点，同一层的节点按照从左到右的顺序打印。


//  Definition for a binary tree node.
class TreeNode {
  public $val = null;
  public $left = null;
  public $right = null;
  function __construct($value) { $this->val = $value; }
}

class Solution {

    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    function levelOrder($root) {
        if($root == null){
            return [];
        }
        $res = [];
        $queue = [$root];
        while (count($queue) > 0){
            $node = $queue[0];
            $queue = array_slice($queue,1);
            $res[] = $node->val;
            if($node->left){
                $queue[] = $node->left;
            }
            if($node->right){
                $queue[] = $node->right;
            }
        }
        return $res;
    }

    // 剑指 Offer 32 - II. 从上到下打印二叉树 II
    function t($root){
        if($root == null){
            return [];
        }
        $res = [];
        $queue = [$root];
        while (count($queue) > 0){
            $tmp = [];
            for($i = count($queue);$i > 0;$i--){
                $node = $queue[0];
                $queue = array_slice($queue,1);
                $tmp[] = $node->val;
                if($node->left){
                    $queue[] = $node->left;
                }
                if($node->right){
                    $queue[] = $node->right;
                }
            }
            $res[] = $tmp;
        }
        return $res;
    }

    function node(){
        $node = new TreeNode(3);
        $node->left = new TreeNode(9);
        $node->right = new TreeNode(20);
        $node->right->left = new TreeNode(15);
        $node->right->right = new TreeNode(7);
        return $node;
    }
}

$obj = new Solution();
$ret = $obj->levelOrder($obj->node());
print_r($ret);

File: go/sword_offer_32_i.go
package main

import (
	"fmt"
)

// 剑指 Offer 32 - I. 从上到下打印二叉树

// 从上到下打印出二叉树的每个节点，同一层的节点按照从左到右的顺序打印。


// 从上至下打印：二叉树的广度优先搜索(BFS)
// BFS通常借助队列的先入先出特性实现


  //Definition for a binary tree node.
type TreeNode struct {
  Val int
  Left *TreeNode
  Right *TreeNode
}

func main()  {
	ret := levelOrder(node())
	fmt.Println(ret)
}

func levelOrder(root *TreeNode) []int {
	var res = make([]int,0)
	if root == nil {
		return res
	}
	// 包含根节点的队列queue
	queue := make([]*TreeNode,0)
	queue = append(queue,root)
	for len(queue) > 0 {
		// 首元素出队
		node := queue[0]
		res = append(res,node.Val)
		queue = queue[1:]
		// 如果左节点有值，则将左节点加入队列
		if node.Left != nil {
			queue = append(queue,node.Left)
		}
		// 如果右节点有值，则将有节点加入队列
		if node.Right != nil {
			queue = append(queue,node.Right)
		}
	}
	return res
}

func node() *TreeNode {
	return &TreeNode{
		Val: 3,
		Left: &TreeNode{
			Val: 9,
		},
		Right: &TreeNode{
			Val: 20,
			Left: &TreeNode{
				Val: 15,
			},
			Right: &TreeNode{
				Val: 7,
			},
		},
	}
}